rasa/core/policies/ted_policy.py

```diff
@@ -923,62 +909,101 @@ def _embed_label(self, label_in: Union[tf.Tensor, np.ndarray]) -> tf.Tensor:
-        name_features = None
-        text_features = None
-        batch_features = []
-        feats_to_combine = [f"{DIALOGUE_FEATURES}_user", f"{DIALOGUE_FEATURES}_action"]
-        for feature in feats_to_combine:
-            if not batch[feature] == []:
-                text_features = self._combine_sparse_dense_features(
-                    batch[feature], feature
-                )
-                text_features = self._tf_layers[f"ffnn.{feature}"](text_features)
-                mask = tf.cast(
-                    tf.math.equal(batch[f"{feature}_if_text"], 1), tf.float32
-                )
-                text_features = text_features * mask
-            if not batch[f"{feature}_name"] == []:
-                name_features = self._combine_sparse_dense_features(
-                    batch[f"{feature}_name"], f"{feature}_name"
-                )
-                name_features = self._tf_layers[f"ffnn.{feature}_name"](name_features)
-                mask = tf.cast(
-                    tf.math.equal(batch[f"{feature}_if_text"], -1), tf.float32
-                )
-                name_features = name_features * mask
+    def _encode_features_per_attribute(
+        self, batch: Dict[Text, Dict[Text, List[tf.Tensor]]], attribute: Text
+    ) -> Optional[tf.Tensor]:
+        if not batch[attribute]:
+            return None
 
-            if text_features is not None and name_features is not None:
-                batch_features.append(text_features + name_features)
-            else:
-                batch_features.append(
-                    text_features if text_features is not None else name_features
-                )
+        mask = batch[attribute][MASK][0]
+        attribute_features = self._combine_sparse_dense_features(
+            batch[attribute][SENTENCE], f"{attribute}_{SENTENCE}", mask=mask
+        )
+
+        if attribute in FEATURES_TO_ENCODE:
+            attribute_features = self._tf_layers[f"ffnn.{attribute}_{SENTENCE}"](
+                attribute_features
+            )
 
-        if not batch[f"{DIALOGUE_FEATURES}_entities"] == []:
-            batch_features.append(batch[f"{DIALOGUE_FEATURES}_entities"])
+        return attribute_features
+
+    def _process_batch(
+        self, batch: Dict[Text, Dict[Text, List[tf.Tensor]]]
+    ) -> tf.Tensor:
+        batch_encoded = {
+            key: self._encode_features_per_attribute(batch, key)
+            for key in batch.keys()
+            if LABEL_KEY not in key and DIALOGUE not in key
+        }
+
+        if (
+            batch_encoded.get(ACTION_TEXT) is not None
+            and batch_encoded.get(ACTION_NAME) is not None
+        ):
+            batch_action = batch_encoded.pop(ACTION_TEXT) + batch_encoded.pop(
+                ACTION_NAME
+            )
+        elif batch_encoded.get(ACTION_TEXT) is not None:
+            batch_action = batch_encoded.pop(ACTION_TEXT)
+        else:
+            batch_action = batch_encoded.pop(ACTION_NAME)
```

Comment:
let's put these `if`s in a helper method, since they repeat below

Reply:
actually is it simpler to use `0` instead of `None` then simple summation should work

Reply:
not 100% sure what you mean

Reply:
if `batch_encoded.pop(ACTION_NAME)` or `batch_encoded.pop(ACTION_TEXT)` is `0` instead of `None` then we don't actually need to have these ifs, we can simply write `batch_action = batch_encoded.pop(ACTION_TEXT) + batch_encoded.pop(ACTION_NAME)`, no?